### PR TITLE
[PERS-26] Template API update

### PR DESCRIPTION
**Pull request by kelvinqian00**
- Update Template API functions
  - Remove `validate-statement-vs-template` functions
  - Change `template->validator` to `compile-templates->validators`
    - Accepts multiple templates; can filter using `:selected-templates`
  - Change `profile->validator` to `compile-profiles->validators`
    - Accepts multiple profiles; can filter using `:selected-profiles` and `:selected-templates`
   - Change `validate-statement-vs-profile` to just `validate-statement`
     - Add multiple public validation helper functions that `validate-statement` calls
     - Rename kwargs: `:option` and `:result` OCaml-isms to `:filter` and `:errors`, respectively
     - Add `:all-valid?` kwarg to specify that ALL templates have to match for the statement
     - Add `:short-circuit?` kwarg to specify that only the first validation error should be returned
- Remove automatic JSON string parsing (**note:** affects Pattern API also)
- Apply Template ID clash check to `compile-patterns->fsms` (**note:** affects Pattern API)
- Update template validation result map
  - `:rule` can now also be `:prop` or `:sref` to distinguish between different types of errors/failures better
  - Change`:determining-property` and `:prop-vals` to `:det-prop` and `:match-vals`, respectively
  - Change `:failure` to `:sref-failure`
 - Move profile, template, and statement asserts ~and coercions~ to util namespaces.
 - Move `profile->id-template-map` and `statement-batch->id-statement-map` convenience fns to the new `template.statement-ref` namespace.
 - Make `?statement-ref-opts` a non-optional (albeit nilable) arg for `template/create-template-predicate` and `tempalte/create-template-validator`.
 - Update `:vals` field in template error map to _always_ be a vector, even for Statement Ref errors
 - Spec fixes:
   - FSMs/Patterns:
     - Update `fsm-spec/valid-transition-src-states?` such that source states only have to be a _subset_, not equals to, the total states.
     - Add `:meta?` to spec for `fsm/plus-nfa`.
     - Correct instrumentation for `read-next` to work with NFAs as well as DFAs.
     - Fix subregistration specs in `utils/statement`.
   - Templates:
     - Let `statement-ref/get-template-fn` and `statement-ref/get-statement-fn` specs to allow for `nil` returns and fix arg generation.
      - Add missing `:every-val-present?` entry to `::template/pred` spec.
     - Fix `validator-spec`, `create-template-validator` and `create-template-predicate` template specs.
   - Persephone API
     - Fix typo in `::persephone/validator-fn` and `::persephone/predicate-fn` spec names.
     - Align specs for `validate-statement-errors` and `validate-statement` (for `:fn-type :errors`).
     - Add missing `::persephone/print?` spec for `match-statement`.
     - Fix bugs relating to the `::persephone/error` spec.

### CODE REVIEW

**Review by FeLungs — COMMENTED** [review]
looks clean / easy to use

**Comment by milt** [comment]
Are there doc updates for this? Would make it easier to adapt stuff like xapipe if the API has breaking changes.

**Comment by kelvinqian00** [comment]
Sorry; that was a copy-paste fail.

File: README.md [hunk 1]
```diff
@@ -35,26 +26,34 @@ Validating a Statement against a Statement Template involves three aspects:
...
+- `:assertion` - Throws an exception if the Statement is invalid, returns `nil` otherwise.
 - `:printer` - Prints an error message when the Statement is invalid. Always returns `nil`.
+- `:templates` - Returns a vector of the IDs of the Statement Templates the Statement is valid for.
 
-`validation-statement-vs-profile` takes an additional option for `:fn-type`:
+`validate-statement` also takes the following two keyword args:
 
-- `:templates` - Returns a vector of the IDs of the Statement Templates the Statement is valid for.
+- `:all-valid?` - If `false` (default), the Statement is considered valid if _any_ of the Statement Template is valid for it. If `true`, validity is if _all_ of the Templates are valid for it. Applicable to all function types except `:templates`.
+- `:short-circuit?` - If `false` (default), returns error data for all invalid Templates; if `true`, returns data for only the first invalid Template found. Applicable to `:result`, `:assertion`, and `:printer`.
```
Comment (milt) [inline README.md:54]: These are really useful, and thoughtful to provide. The combination of the two gives lib consumers a lot of flexibility

**Comment by milt** [comment]
> Remove JSON string parsing (note: affects Pattern API also)

OK, how does a lib consumer deserialize a profile now? Just vanilla JSON parse?

**Comment by kelvinqian00** [comment]
> OK, how does a lib consumer deserialize a profile now? Just vanilla JSON parse?

As per your recommendations, they would need to have the Statements and Profiles parsed into EDN _before_ passing them into the API. They would do regular JSON parsing, just keywording keys for Profiles and stringifying keys for Statements.

**Comment by milt** [comment]
> As per your recommendations, they would need to have the Statements and Profiles parsed into EDN before passing them into the API. They would do regular JSON parsing, just keywording keys for Profiles and stringifying keys for Statements.

That sounds good. `#(keyword nil %)` works for all cases, right? Like there's no part of profiles with IRI keys?

**Comment by kelvinqian00** [comment]
> That sounds good. #(keyword nil %) works for all cases, right? Like there's no part of profiles with IRI keys?

No (at least not in the spec, though I guess the user can add extraneous IRI keys, but then that's their problem). The main troublesome key would be `@context`, but that gets ignored by Persephone post-validation so they could just convert it to `:_context`.

**Comment by milt** [comment]
> ...so they could just convert it to `:_context`

So you're saying the lib consumer _will_ have to do special key coercion then, right?

**Comment by kelvinqian00** [comment]
> So you're saying the lib consumer will have to do special key coercion then, right?

Erm yes, but that's more of a Pan problem than a Persephone problem. But I see your point - maybe add back the `coerce-profile` and `coerce-statement` functions as functions the user can call?

**Comment by milt** [comment]
> ...maybe add back...

Yeah that sounds like a good balance, so you don't have to force a parsing choice but folks can have guarantees about structure

**Review by milt — CHANGES_REQUESTED** [review]
~~Looks like there are some additional spec/instrumentation issues that come up when using the new API for direct template compilation. See concept filter test fails on https://github.com/yetanalytics/xapipe/pull/68.~~

It appears there is actually a bug in DATASIM or xapipe's test profile that is yielding invalid statements, it doesn't look like persephone.

**Review by milt — APPROVED** [review]
OK, all the downstream issues are outside of persephone's purview, this is looking really good.

**Review by cliffcaseyyet — APPROVED** [review]

**Pull request merged**